introduce `pulumi package info`

tgummerer opened the pull request:
Introduce a new `pulumi package info` command that allows the user to get details of a package without having to read through the whole schema.  It filters out some details, and intentionally simplifies some of it to make it more accessible.

Examples:
```
$ pulumi package info aws
Provider: aws (6.75.0)
  Description: A Pulumi package for creating and managing Amazon Web Services (AWS) cloud resources.
Total resources: 1502
Total modules: 214

Use 'pulumi package info --module <module> aws' to list resources in a module
Use 'pulumi package info --module <module> --resource <resource> aws' for detailed resource info
```

```
$ pulumi package info aws --module s3
Module: aws:s3
Resources: 27

 - AccessPoint: Provides a resource to manage an S3 Access Point.
 - AccountPublicAccessBlock: Manages S3 account-level Public Access Block configuration. For more information about these settings, see the [AWS S3 Block Public Access documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/access-control-block-public-access.html).
 - AnalyticsConfiguration: Provides a S3 bucket [analytics configuration](https://docs.aws.amazon.com/AmazonS3/latest/dev/analytics-storage-class.html) resource.
 - Bucket: Provides a S3 bucket resource.
 - BucketAccelerateConfigurationV2: Provides an S3 bucket accelerate configuration resource. See the [Requirements for using Transfer Acceleration](https://docs.aws.amazon.com/AmazonS3/latest/userguide/transfer-acceleration.html#transfer-acceleration-requirements) for more details.
 - BucketAclV2: Provides an S3 bucket ACL resource.
 - BucketCorsConfigurationV2: Provides an S3 bucket CORS configuration resource. For more information about CORS, go to [Enabling Cross-Origin Resource Sharing](https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html) in the Amazon S3 User Guide.
[....]
```

```
$ pulumi package info aws --module s3 --resource Bucket
Resource: aws:s3:Bucket

Input Properties:
 - accelerationStatus (string (optional)): Sets the accelerate configuration of an existing bucket. Can be `Enabled` or `Suspended`.
 - acl (string (optional)): The [canned ACL](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl) to apply. Valid values are `private`, `public-read`, `public-read-write`, `aws-exec-read`, `authenticated-read`, and `log-delivery-write`. Defaults to `private`.  Conflicts with `grant`.
 - arn (string (optional)): The ARN of the bucket. Will be of format `arn:aws:s3:::bucketname`.
 - bucket (string (optional)): The name of the bucket. If omitted, this provider will assign a random, unique name. Must be lowercase and less than or equal to 63 characters in length. A full list of bucket naming rules [may be found here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html).
 - bucketPrefix (string (optional)): Creates a unique bucket name beginning with the specified prefix. Conflicts with `bucket`. Must be lowercase and less than or equal to 37 characters in length. A full list of bucket naming rules [may be found here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html).
 - corsRules ([]BucketCorsRule (optional)): A rule of [Cross-Origin Resource Sharing](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) (documented below).
[...]

Output Properties:
 - accelerationStatus (string (always present)): Sets the accelerate configuration of an existing bucket. Can be `Enabled` or `Suspended`.
 - acl (string): The [canned ACL](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl) to apply. Valid values are `private`, `public-read`, `public-read-write`, `aws-exec-read`, `authenticated-read`, and `log-delivery-write`. Defaults to `private`.  Conflicts with `grant`.
 - arn (string (always present)): The ARN of the bucket. Will be of format `arn:aws:s3:::bucketname`.
 - bucket (string (always present)): The name of the bucket. If omitted, this provider will assign a random, unique name. Must be lowercase and less than or equal to 63 characters in length. A full list of bucket naming rules [may be found here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html).
[...]
```
Fixes https://github.com/pulumi/pulumi/issues/19097

mikhailshilkov commented:
Could you please extend the PR description to show a few examples of the command usage and output?

tgummerer commented:
There's some examples in the tests, but I've added some real world ones to the PR description now.

Frassle commented on pkg/cmd/pulumi/packagecmd/package_info.go:
```diff
@@ -0,0 +1,283 @@
+		}
+		return prop.Ref, nil
+	}
+	if prop.Ref != "" {
```
Duplicate of the above check?

tgummerer replied:
Yeah indeed, we never get here.  Removed.

Frassle commented on pkg/cmd/pulumi/packagecmd/package_info.go:
```diff
@@ -0,0 +1,283 @@
+
+func getType(spec *schema.PackageSpec, prop schema.TypeSpec) (string, error) {
+	typ := prop.Type
+	if typ != "" && typ != "object" && typ != "array" {
```
Need to check for "ref" here?

tgummerer replied:
I don't think so, since `Type` shouldn't be set in that case.  I added a check for `&& prop.Ref == ""` now though.

Frassle commented on pkg/cmd/pulumi/packagecmd/package_info.go:
```diff
@@ -0,0 +1,283 @@
+	if prop.Ref != "" {
+		return prop.Ref, nil
+	}
+	return "unknown", nil
```
No handling for objects/maps?

tgummerer replied:
Fixed now.

Frassle commented on pkg/cmd/pulumi/packagecmd/package_info_test.go:
```diff
@@ -0,0 +1,186 @@
+						"enumProp": {
+							Description: "this is an enum property",
+							TypeSpec: schema.TypeSpec{
+								Ref: "#/types/test:index:EnumType",
```
This should have Type: "ref" I think

tgummerer replied:
I don't think so.  It is documented as being ignored in the schema: https://github.com/pulumi/pulumi/blob/tg%2Fpackage-info/pkg/codegen/schema/pulumi.json#L302.  

From a quick look at a couple of schemas (aws/awsx) don't have the type set at all when there's a ref specified.

mikhailshilkov commented:
The command works, we probably want to try to make it nice too. I've shared my high-level feedback in https://docs.google.com/document/d/1AzXPYoLml2FMs1zqxZo2M7nM052QefJdf5n4j8dxFQA/edit?tab=t.0

There are probably other opportunities in the same theme of making it easier to use in different scenarios, including to debug your local packages.

Frassle commented on pkg/cmd/pulumi/packagecmd/package_info_test.go:
```diff
@@ -0,0 +1,194 @@
+\x1b[1mOutputs\x1b[0m:
+(All input properties are implicitly available as output properties)
+ - \x1b[1marrayProp\x1b[0m (\x1b[4m[]TestType\x1b[0m\x1b[4m\x1b[0m): this is an array property
+ - \x1b[1menumProp\x1b[0m (\x1b[4menum(string){EnumValue1, value2}\x1b[0m\x1b[4m\x1b[0m): this is an enum property
```
This should have object and map items in the test as well, also a ref to a non-enum type

tgummerer replied:
Isn't object and map the same?  We have a ref to a non-enum type above, with `[]TestType`.  I've added the object now.

Frassle replied:
objects have Properties, maps have AdditionalProperties

julienp commented on pkg/cmd/pulumi/packagecmd/package_info.go:
```diff
@@ -0,0 +1,382 @@
+}
+
+func summaryFromDescription(description string) string {
+	// The description of a resource is markdown formatted.  We only want to provide aa
```
```suggestion
	// The description of a resource is markdown formatted.  We only want to provide a
```

julienp commented:
I am not getting the module name when running with `--module` (possibly intended, but the PR description shows it)

```
pulumi % pulumi package info aws --module s3
Name: aws
Version: 6.77.0
Description: A Pulumi package for creating and managing Amazon Web Services (AWS) cloud resources.
Resources: 27

 - AccessPoint: Provides a resource to manage an S3 Access Point.
...
```

julienp commented:
I assume the simplified module handling so that we don't have each resource in its own module for bridged providers, because the modules are things like `index/randomBytes` or `s3/bucket`, and we want to group on `index` and `s3` in these cases. I like that, but I think we should still allow looking up by the real module name.

```
% pulumi package info random --module index/randomBytes
...
error: module "index/randomBytes" not found
% pulumi package info random -r RandomBytes
Resource: random:index/randomBytes:RandomBytes
...
```

tgummerer commented:
> I am not getting the module name when running with --module (possibly intended, but the PR description shows it)

Yeah it should be there.  I've addressed this in the latest commit.

>  I like that, but I think we should still allow looking up by the real module name.

Agreed, and fixed in the latest commit as well.
